mirrord/agent/src/steal/connection.rs

```diff
@@ -591,28 +604,131 @@ impl TcpConnectionStealer {
+        }
+
+        // Headers.
+        for (name, value) in headers.into_iter() {
+            if let Some(name) = name {
+                bytes.put(&b"\r\n"[..]);
+                bytes.put(name.as_str().as_bytes());
+                bytes.put(&b": "[..]);
+            } else {
+                // TODO: are multiple values for the same header comma delimited?
+                bytes.put(&b","[..])
+            }
+            bytes.put(value.as_bytes())
+        }
+        bytes.put(&b"\r\n\r\n"[..]);
+
+        // Body.
+        bytes.put(body.collect().await.unwrap().to_bytes()); // Unwrap Infallible.
+
+        bytes.to_vec()
+    }
+
+    fn handle_response_reconstruction_fail(err: hyper::http::Error) -> hyper::http::Error {
+        error!("Could not reconstruct http response: {err:?}");
+        // TODO: What should happen? Currently just skipping that response, which would
+        //       also spoils the rest of the stream. Should hopefully never happen.
+        debug_assert!(false);
+        err
+    }
+
+    /// Forward an HTTP response to a stolen HTTP request from the layer to the back to the browser.
     ///
-    ///                       _________________________agent_________________________
-    /// Local App -> Layer -> ClientConnectionHandler -> Stealer -> HttpFilterManager -> Browser
-    ///                                                          ^- You are here.
-    async fn http_response(
-        &mut self,
-        response: HttpResponse,
-    ) -> std::result::Result<(), AgentError> {
-        todo!();
-        match self.port_subscriptions.get(&response.port) {
-            Some(HttpFiltered(manager)) => todo!(),
-            Some(Unfiltered(_)) | None => {
-                warn!(
-                    "Got an http response to port {:?} which is not being filtered. \
-                Maybe the layer unsubscribed and this is a remainder of a closed connection. \
-                Not forwarding.",
-                    response.port
-                );
-                Ok(())
+    ///                         _______________agent_______________
+    /// Local App --> Layer --> ClientConnectionHandler --> Stealer --> Browser
+    ///                                                             ^- You are here.
+    async fn http_response(&mut self, response: HttpResponse) -> Result<()> {
+        let tcp_stream_write_half;
+        if let Some(stream) = self.http_write_streams.get_mut(&response.connection_id) {
+            tcp_stream_write_half = stream;
+        } else {
+            warn!(
+                "Got an http response in a connection for which no tcp stream is present. Not forwarding.",
+            );
+            return Ok(());
+        }
+        let counter = self
+            .http_request_counters
+            .entry(response.connection_id)
+            .or_insert(0);
+        if response.request_id == *counter {
+            if let Ok(response) = response
+                .response
```

Comment:
What's response.response?

Reply:
It's the `InternalHttpResponse` of the `HttpResponse`. Should that member be renamed to `internal_response`?